From 8b9742d9f7d19dc68247356bb1bd318456c3f3a5 Mon Sep 17 00:00:00 2001
From: mvulsx <mvuls@maxlinear.com>
Date: Tue, 4 Jul 2023 09:59:14 +0000
Subject: [PATCH] WLANRTSYS-56685: wpa_supplicant: extend vendor_elem params to
 cover Auth Req frames

The new VENDOR_ELEM value 15 can now be used to add a vendor element
into Authentication Request frames.

For example:
VENDOR_ELEM_ADD 15 dd05001122330a
and to clear that:
VENDOR_ELEM_REMOVE 15 *

Signed-off-by: mvulsx <mvuls@maxlinear.com>
---
 src/common/wpa_ctrl.h |  1 +
 wpa_supplicant/sme.c  | 24 ++++++++++++++++++++++++
 2 files changed, 25 insertions(+)

diff --git a/src/common/wpa_ctrl.h b/src/common/wpa_ctrl.h
index 80b0c6bb2..d96579444 100644
--- a/src/common/wpa_ctrl.h
+++ b/src/common/wpa_ctrl.h
@@ -523,6 +523,7 @@ enum wpa_vendor_elem_frame {
 	VENDOR_ELEM_P2P_ASSOC_RESP = 12,
 	VENDOR_ELEM_ASSOC_REQ = 13,
 	VENDOR_ELEM_PROBE_REQ = 14,
+	VENDOR_ELEM_AUTH_REQ = 15,
 	NUM_VENDOR_ELEM_FRAMES
 };
 
diff --git a/wpa_supplicant/sme.c b/wpa_supplicant/sme.c
index 2a9ffa0c2..ce6059260 100644
--- a/wpa_supplicant/sme.c
+++ b/wpa_supplicant/sme.c
@@ -1127,6 +1127,30 @@ no_fils:
 
 	wpa_s->sme.wds = wpa_s->conf->wds;
 
+	if (wpa_s->vendor_elem[VENDOR_ELEM_AUTH_REQ]) {
+		struct wpabuf *buf = wpa_s->vendor_elem[VENDOR_ELEM_AUTH_REQ];
+		errno_t err;
+		u8 auth_vs_ie[128];
+
+		if (params.ie) {
+			err = memcpy_s((void *)params.ie + params.ie_len, wpabuf_len(buf), wpabuf_head(buf), wpabuf_len(buf));
+			if (EOK != err) {
+				wpa_printf(MSG_ERROR, "%s: memcpy_s error %d", __func__, err);
+				wpabuf_free(resp);
+				return;
+			}
+		} else  {
+			err = memcpy_s(auth_vs_ie, sizeof(auth_vs_ie), wpabuf_head(buf), wpabuf_len(buf));
+			if (EOK != err) {
+				wpa_printf(MSG_ERROR, "%s: memcpy_s error %d", __func__, err);
+				wpabuf_free(resp);
+				return;
+			}
+			params.ie = auth_vs_ie;
+		}
+		params.ie_len += wpabuf_len(buf);
+	}
+
 	wpa_supplicant_cancel_sched_scan(wpa_s);
 	wpa_supplicant_cancel_scan(wpa_s);
 
-- 
2.43.0

